File: Fun/Chess/ChessPiece.java
package Chess;

public abstract class ChessPiece implements Cloneable {
    protected char color; // 'w' or 'b'

    public ChessPiece(char color) {
        this.color = color;
    }

    public char getColor() {
        return color;
    }

    public abstract boolean isValidMove(int fromRow, int fromCol, int toRow, int toCol, ChessBoard board);

    @Override
    public ChessPiece clone() {
        try {
            return (ChessPiece) super.clone();
        } catch (CloneNotSupportedException e) {
            throw new AssertionError();
        }
    }

    @Override
    public abstract String toString();

    public abstract String getType();
    public abstract int getValue();
}

File: site/public/Chess-js/ChessPiece.js
class ChessPiece {
    constructor(color) {
        this.color = color; // 'w' or 'b'
    }

    getColor() {
        return this.color;
    }

    isValidMove(fromRow, fromCol, toRow, toCol, chessBoard) {
        throw new Error("isValidMove must be implemented by subclass");
    }

    clone() {
        return Object.assign(Object.create(Object.getPrototypeOf(this)), this);
    }

    toString() {
        throw new Error("toString must be implemented by subclass");
    }

    getType() {
        throw new Error("getType must be implemented by subclass");
    }

    getValue() {
        throw new Error("getValue must be implemented by subclass");
    }
}